10/WAKU2: General improvements to main spec (#444)
vacp2p/rfc: content/docs/rfcs/10/README.md

This makes Waku ideal for running a p2p protocol on mobile and in similarly restricted environments.

Historically, it has its roots in [Waku v1](/spec/6),
which stems from [Whisper](https://eips.ethereum.org/EIPS/eip-627), originally part of the Ethereum stack.
However, Waku v2 acts more as a thin wrapper for PubSub and has a different API.
It is implemented in an iterative manner where initial focus is on porting essential functionality to libp2p.
See [rough road map (2020)](https://vac.dev/waku-v2-plan) for more historical context.

## Motivation and goals

Waku as a family of protocols is designed to have a set of properties that are useful for many applications:

1. **Useful for generalized messaging.**

Many applications require some form of messaging protocol to communicate between different subsystems or different nodes.
This messaging can be human-to-human or machine-to-machine or a mix.
Waku is designed to work for all these scenarios.

2. **Peer-to-peer.**

Applications sometimes have requirements that make them suitable for peer-to-peer solutions:
- Censorship-resistant with no single point of failure 
- Adaptive and scalable network
- Shared infrastructure

> 3. **Runs anywhere**.

Reviewer: ```suggestion
3. **Runs anywhere.**
```

Applications often run in restricted environments, where resources or the environment is restricted in some fashion.
For example:

- limited bandwidth, CPU, memory, disk, battery, etc
- not being publicly connectable

> - only being intermittently connected; mostly-offline

Reviewer: ```suggestion
- Limited bandwidth, CPU, memory, disk, battery, etc
- Not being publicly connectable
- Only being intermittently connected; mostly-offline
```

4. **Privacy-preserving.**

 Applications often have a desire for some privacy guarantees, such as:
 - Pseudonymity and not being tied to any personally identifiable information (PII)
 - Metadata protection in transit
 - Various forms of unlinkability, etc

5. **Modular design.**

Applications often have different trade-offs when it comes to what properties they and their users value.
Waku is designed in a modular fashion where an application protocol or node can choose what protocols they run.
We call this concept *adaptive nodes*.

For example:
- resource usage vs metadata protection
- providing useful services to the network vs mostly using it

> - stronger guarantees for spam protection vs economic registration cost

Reviewer: ```suggestion
- Resource usage vs metadata protection
- Providing useful services to the network vs mostly using it
- Stronger guarantees for spam protection vs economic registration cost
```

## Network interaction domains

> While Waku is best though of as a single cohesive thing, there are three network interaction domains:

Reviewer: ```suggestion
While Waku is best thought of as a single cohesive thing, there are three network interaction domains:
```

(a) gossip domain
(b) discovery domain
(c) req/resp domain

### Protocols and identifiers

Since Waku v2 is built on top of libp2p, many protocols have a libp2p protocol identifier.
The current main [protocol identifiers](https://docs.libp2p.io/concepts/protocols/) are:

1. `/vac/waku/relay/2.0.0`
2. `/vac/waku/store/2.0.0-beta3`
3. `/vac/waku/filter/2.0.0-beta1`
4. `/vac/waku/lightpush/2.0.0-beta1`

> This is in addition to protocols that specify mesages, payloads, and recommend usages.

Reviewer 2: ```suggestion
This is in addition to protocols that specify messages, payloads, and recommended usages.
```

for liveness checks between peers, or to keep peer-to-peer connections alive.

> For the actual content being passed around, see the [14/WAKU2-MESSAGE](/spec/14) and [26/WAKU2-PAYLOAD](/spec/26).

Reviewer: if I were new to specs then "the actual content being passed around," would sound a bit confusing to me. 
How do we define what is the actual content? also, what do we mean by "being passed around", e.g., in store protocol, we pass history queries, and in response, we get the list of waku messages, in this case, what is the actual content that is being passed?

Author: Good point, this should be clarified

Author: decided to simplify and remove this, already mentioned elsewhere

- [go-waku (Go)](https://github.com/status-im/go-waku/)
- [js-waku (NodeJS and Browser)](https://github.com/status-im/js-waku/)

> Below you can find an overview of the specs that they implement as they relate to Waku.

Reviewer: ```suggestion
Below you can find an overview of the specs that they implement as they relate to Waku v2.
```

In general, I suggest using "Waku v2" instead of "Waku" as much as possible.
